src/smtml/expr.ml

```diff
@@ -320,13 +303,66 @@ let exists vars body = binder Exists vars body
 
 let raw_unop ty op hte = make (Unop (ty, op, hte)) [@@inline]
 
+let is_val_real e =
+  match view e with
+  | Val (Real x) -> Float.is_finite x
+  | _ -> false
+
+let normalize_eq_or_ne op (ty', e1, e2) =
+  let is_real = is_val_real e1 && is_val_real e2 in
+  let make_relop lhs rhs = Relop (ty', op, lhs, rhs) in
+  let ty,ty2 = ty e1,ty e2 in
+  if Ty.equal ty ty2 then
+  match ty with
+  | Ty_bitv a ->
+     let binop = make (Binop (ty, Sub, e1, e2)) in
+     let zero = if a=32 then make (Val (Bitv (Bitvector.of_int32 0l)))
+                else make (Val (Bitv (Bitvector.of_int64 0L)))
+     in
+      make_relop binop zero
+  | Ty_fp 32 ->
+      let lhs, rhs = if is_real then
+          make (Binop (ty, Sub, e1, e2)), make (Val (Num (F32 0l)))
+        else e1, e2
+      in
+      make_relop lhs rhs
+  | Ty_fp 64 ->
+      let lhs, rhs = if is_real then
+          make (Binop (ty, Sub, e1, e2)), make (Val (Num (F64 0L)))
+        else e1, e2
+      in
+      make_relop lhs rhs
```

Comment:
From what I understand, you're using `is_real` to check whether `Val (Real x)` satisfies `Float.is_finite x`. In the case where the type is `Ty_fp _`, you can never have a numeric `Val (Real _)`, so the `if` statement is redundant, it will always take the false branch. A more appropriate pattern match would be the following:


```suggestion
  | Ty_real -> make (Binop (ty, Sub, e1, e2)), make (Val (Real 0.)))
```

This only applies to arithmetic reals, and it's safe to remove the `is_real` check because real arithmetic does not include special floating-point values like `inf`.